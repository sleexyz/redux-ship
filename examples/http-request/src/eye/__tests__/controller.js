// @flow
import 'babel-polyfill';
import {createStore} from 'redux';
import * as Ship from 'redux-ship';
import * as Effect from '../../effect';
import * as EyeController from '../controller';
import * as EyeModel from '../model';

const liveSnapshot =
  {"action":{"type":"Load"},"snapshot":[{"type":"Commit","commit":{"type":"LoadStart"}},{"type":"Effect","effect":{"type":"HttpRequest","url":"http://swapi.co/api/people/3/"},"result":"{\"name\":\"R2-D2\",\"height\":\"96\",\"mass\":\"32\",\"hair_color\":\"n/a\",\"skin_color\":\"white, blue\",\"eye_color\":\"red\",\"birth_year\":\"33BBY\",\"gender\":\"n/a\",\"homeworld\":\"http://swapi.co/api/planets/8/\",\"films\":[\"http://swapi.co/api/films/5/\",\"http://swapi.co/api/films/4/\",\"http://swapi.co/api/films/6/\",\"http://swapi.co/api/films/3/\",\"http://swapi.co/api/films/2/\",\"http://swapi.co/api/films/1/\",\"http://swapi.co/api/films/7/\"],\"species\":[\"http://swapi.co/api/species/2/\"],\"vehicles\":[],\"starships\":[],\"created\":\"2014-12-10T15:11:50.376000Z\",\"edited\":\"2014-12-20T21:17:50.311000Z\",\"url\":\"http://swapi.co/api/people/3/\"}"},{"type":"Commit","commit":{"type":"LoadSuccess","color":"red"}}]};

test('controller with live snapshot', () => {
  const {action, snapshot} = liveSnapshot;
  expect(Ship.simulate(EyeController.control(action), snapshot)).toEqual(snapshot);
});

function runEffect(effect: Effect.Effect): any {
  switch (effect.type) {
  case 'HttpRequest':
    return JSON.stringify({eye_color: 'red'});
  default:
    return;
  }
}

test('controller with generated snapshot', async () => {
  const action = {type: 'Load'};
  const store = createStore(EyeModel.reduce, EyeModel.initialState);
  const ship = Ship.snap(EyeController.control(action));
  const snapshot = await Ship.run(runEffect, store, ship);
  expect(snapshot).toMatchSnapshot();
});

test('controller step by step', () => {
  const action = {type: 'Load'};
  const gen = EyeController.control(action);
  expect(gen.next()).toMatchSnapshot();
  expect(gen.next()).toMatchSnapshot();
  expect(gen.next(JSON.stringify({eye_color: 'red'}))).toMatchSnapshot();
  expect(gen.next()).toMatchSnapshot();
});
