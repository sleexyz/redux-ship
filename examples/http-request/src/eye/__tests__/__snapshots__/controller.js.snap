exports[`test controller step by step 1`] = `
Object {
  "done": false,
  "value": Object {
    "command": Object {
      "commit": Object {
        "type": "LoadStart",
      },
      "type": "Commit",
    },
    "type": "Command",
  },
}
`;

exports[`test controller step by step 2`] = `
Object {
  "done": false,
  "value": Object {
    "command": Object {
      "effect": Object {
        "type": "HttpRequest",
        "url": "http://swapi.co/api/people/3/",
      },
      "type": "Effect",
    },
    "type": "Command",
  },
}
`;

exports[`test controller step by step 3`] = `
Object {
  "done": false,
  "value": Object {
    "command": Object {
      "commit": Object {
        "color": "red",
        "type": "LoadSuccess",
      },
      "type": "Commit",
    },
    "type": "Command",
  },
}
`;

exports[`test controller step by step 4`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`test controller with generated snapshot 1`] = `
Object {
  "result": undefined,
  "snapshot": Array [
    Object {
      "commit": Object {
        "type": "LoadStart",
      },
      "type": "Commit",
    },
    Object {
      "effect": Object {
        "type": "HttpRequest",
        "url": "http://swapi.co/api/people/3/",
      },
      "result": "{\"eye_color\":\"red\"}",
      "type": "Effect",
    },
    Object {
      "commit": Object {
        "color": "red",
        "type": "LoadSuccess",
      },
      "type": "Commit",
    },
  ],
}
`;
